src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs

```diff
@@ -0,0 +1,398 @@
+                           else
+                                Seq.empty
+        let dumpClasses = if generateDump 
+                          then seq { dumpMachineClass; dumpRegisterClass }
+                          else Seq.empty
+        Seq.append qubitClasses dumpClasses
+
+    let implMethodName (sp : QsSpecialization) =
+        "Do" + sp.Parent.Name.Value + match sp.Kind with
+                                      | QsBody              -> ""
+                                      | QsAdjoint           -> "Adj"
+                                      | QsControlled        -> "Ctl"
+                                      | QsControlledAdjoint -> "CtlAdj"
+
+    let fullArgsForSpecialization args (sp : QsSpecialization) =
+        match sp.Kind with
+        | QsControlled | QsControlledAdjoint -> ("__ctrls__", "IQArray<Qubit>") :: args
+        | _ -> args
+
+    let buildImplMethod context args (sp : QsSpecialization) =
+        let outType = match sp.Signature.ReturnType.Resolution with
+                      | UnitType -> "void"
+                      | _ -> SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        match sp.Implementation with
+        | Generated SelfInverse -> Seq.empty
+        | _ ->
+            let implName = implMethodName sp
+            let fullArgs = fullArgsForSpecialization args sp
+            let parms = fullArgs |> List.map (fun (n, t) -> ``param`` n ``of`` (``type`` t))
+            let impl = ``method`` outType implName ``<<`` [] ``>>`` 
+                            ``(`` parms ``)`` [``public``;``abstract``]
+                            ``{`` [] ``}`` :> MemberDeclarationSyntax
+            seq { impl }
+
+    let buildImplProperty targetClass context args (sp : QsSpecialization) =
+        let inType  = SimulationCode.roslynTypeName context sp.Signature.ArgumentType
+        let outType = SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        let propertyType = "Func<" + inType + ", " + outType + ">"
+        let propertyName =
+            match sp.Kind with
+            | QsBody              -> "Body"
+            | QsAdjoint           -> "AdjointBody"
+            | QsControlled        -> "ControlledBody"
+            | QsControlledAdjoint -> "ControlledAdjointBody"
+        let propertyFromBody body =
+            (``property-arrow_get`` propertyType propertyName [``public``; ``override``]
+                    ``get`` (``=>`` body)) :> MemberDeclarationSyntax
+        match sp.Implementation with
+        | SpecializationImplementation.Intrinsic -> 
+            let returnType  = sp.Signature.ReturnType
+            let inData = ``ident`` "__in__"
+            let spArgs = match sp.Kind with 
+                         | QsControlled | QsControlledAdjoint -> 
+                            let controls = 
+                                { 
+                                    VariableName = NonNullable<string>.New "__ctrls__" |> ValidName
```

Comment:
For this logic you could actually use the SyntaxGenerator.ImmutableQubitArrayWithName. 
nit: I believe the name needs to match the one in fullArgsForSpecialization? If so, I suggest to have the used string in one place and use that in both functions.

Reply:
The SyntaxGenerator function creates a TYpedExpression, where here I need a LocalVariable, so it doesn't quite work.
The two strings don't need to match, one is used defining the method and the other is used calling the method. I've eliminated other magic strings where they need to match, but since these don't need to match, I don't want the code to force them to match.